File: src/main/repos.ts
import { stat } from "node:fs/promises";
import { basename, dirname } from "node:path";
import type { AddRepoFailure, AddReposResult } from "@shared/types";
import { detectMainBranch, resolveRepoRoot } from "./git";
import * as store from "./store";

// MARK: Failure messages

/**
 * Turn whatever `resolveRepoRoot`/`detectMainBranch` threw into one short line
 * for the UI. git's own text is kept for anything unrecognised, minus the
 * `git <args> failed:` prefix `runGit` adds and any trailing hint lines.
 */
export function describeAddFailure(error: unknown): string {
  const raw = error instanceof Error ? error.message : String(error);
  if (/not a git repository/i.test(raw)) return "Not a git repository";
  if (/ENOENT|ENOTDIR/.test(raw)) return "Folder not found";
  const withoutPrefix = raw.replace(/^git .*? failed: /, "");
  const firstLine = withoutPrefix.split("\n")[0]!.trim();
  return firstLine.replace(/^fatal: /, "") || "Could not add this folder";
}

/**
 * git needs a directory to run in, and a Dock drop can hand us a file (dragging
 * a repo's README, say). Fall back to the containing folder; an unreadable path
 * is passed through so git reports it.
 */
async function containingDirectory(path: string): Promise<string> {
  try {
    return (await stat(path)).isDirectory() ? path : dirname(path);
  } catch {
    return path;
  }
}

// MARK: Adding

/**
 * Add repositories by path — used by both the picker and Dock drops.
 *
 * Best effort per path: each is resolved to its repository root, gets its main
 * branch auto-detected, and is appended to the store. A path that isn't a repo
 * (or is already configured) lands in `failed` instead of aborting the batch.
 * Sequential on purpose: `store.addRepo` is read-modify-write, and duplicates
 * within one batch must be caught.
 */
export async function addRepos(paths: string[]): Promise<AddReposResult> {
  const added: string[] = [];
  const failed: AddRepoFailure[] = [];

  for (const path of paths) {
    try {
      const root = await resolveRepoRoot(await containingDirectory(path));
      if (store.getConfig().repos.some((r) => r.path === root)) {
        failed.push({ path, message: "Already added" });
        continue;
      }
      const mainBranch = await detectMainBranch(root);
      const config = store.addRepo({ path: root, mainBranch });
      // Report the stored (sanitized) display name, not the raw dropped path.
      added.push(config.repos.find((r) => r.path === root)?.name ?? basename(root));
    } catch (error) {
      failed.push({ path, message: describeAddFailure(error) });
    }
  }

  return { config: store.getConfig(), added, failed };
}

File: crates/wtm-core/src/repos.rs
//! Adding repositories by path.

use std::path::{Path, PathBuf};

use crate::git::{detect_main_branch, resolve_repo_root, GitError};

/// Turn whatever `resolve_repo_root`/`detect_main_branch` failed with into one
/// short line for the UI.
pub fn describe_add_failure(err: &GitError) -> String {
    let raw = &err.message;
    let lower = raw.to_ascii_lowercase();
    if lower.contains("not a git repository") {
        return "Not a git repository".to_string();
    }
    if raw.contains("ENOENT") || raw.contains("ENOTDIR") || lower.contains("no such file") {
        return "Folder not found".to_string();
    }
    let without_prefix = match raw.find(" failed: ") {
        Some(i) if raw.starts_with("git ") => &raw[i + " failed: ".len()..],
        _ => raw.as_str(),
    };
    let first = without_prefix.lines().next().unwrap_or("").trim();
    let first = first.strip_prefix("fatal: ").unwrap_or(first);
    if first.is_empty() {
        "Could not add this folder".to_string()
    } else {
        first.to_string()
    }
}

/// git needs a directory to run in, and a drop can hand us a file. Fall back
/// to the containing folder; an unreadable path is passed through so git
/// reports it.
pub async fn containing_directory(path: &Path) -> PathBuf {
    match tokio::fs::metadata(path).await {
        Ok(m) if !m.is_dir() => path
            .parent()
            .map(Path::to_path_buf)
            .unwrap_or_else(|| path.to_path_buf()),
        _ => path.to_path_buf(),
    }
}

/// Resolve a picked path to `(repo root, main branch)`.
pub async fn inspect_repo(path: &Path) -> Result<(String, String), GitError> {
    let dir = containing_directory(path).await;
    let root = resolve_repo_root(&dir).await?;
    let main = detect_main_branch(&root).await;
    Ok((root, main))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn git_error(stderr: &str) -> GitError {
        GitError {
            message: format!("git rev-parse --show-toplevel failed: {stderr}"),
            cwd: "/tmp".into(),
            args: vec![],
            stderr: stderr.into(),
        }
    }

    #[test]
    fn recognises_non_repository() {
        let e = git_error("fatal: not a git repository (or any of the parent directories): .git");
        assert_eq!(describe_add_failure(&e), "Not a git repository");
    }

    #[test]
    fn recognises_missing_folder() {
        let e = GitError {
            message: "spawn git ENOENT".into(),
            cwd: "".into(),
            args: vec![],
            stderr: "".into(),
        };
        assert_eq!(describe_add_failure(&e), "Folder not found");
    }

    #[test]
    fn keeps_gits_wording_without_prefix() {
        let e = git_error(
            "fatal: detected dubious ownership in repository at '/srv/app'\nTo add an\nexception",
        );
        assert_eq!(
            describe_add_failure(&e),
            "detected dubious ownership in repository at '/srv/app'"
        );
    }

    #[test]
    fn falls_back_for_empty() {
        let e = GitError {
            message: "".into(),
            cwd: "".into(),
            args: vec![],
            stderr: "".into(),
        };
        assert_eq!(describe_add_failure(&e), "Could not add this folder");
    }
}
